C++: Constant type bounds in the new range analysis

MathiasVP opened the pull request:
This PR adds type-based bounds in the new range analysis library. This allows us to deduce that fewer things overflow, and thus allows us exclude fewer bounds.

geoffw0 commented on cpp/ql/test/library-tests/ir/range-analysis/SimpleRangeAnalysis_tests.cpp:
```diff
@@ -936,7 +936,7 @@ void two_bounds_from_one_test(short ss, unsigned short us) {
   }
 
-  if (ss + 1 < sizeof(int)) {  // $ overflow=+
+  if (ss + 1 < sizeof(int)) { // $ overflow=-
```
:+1: verified experimentally

MathiasVP commented:
I've verified the results and they LGTM. I'm still a bit unsure about whether we should suppress constant bounds arising purely from types such as:
```cpp
void f(unsigned int ui) {
  unsigned long long ull = ui;
  range(ull); // we infer that `ull <= UINT_MAX` here. Is that what we want?
}
```
In any case, I think we can push this feature to a future PR.

jketema commented on cpp/ql/lib/semmle/code/cpp/rangeanalysis/new/internal/semantic/SemanticExprSpecific.qll:
```diff
@@ -434,6 +434,50 @@ module SemanticExprConfig {
+      integralType = t and limit = 2.pow(8 * integralType.getByteSize())
+    |
+      if integralType instanceof SemBooleanType
+      then lb = 0 and ub = 1
```
I think, technically, this should be `0` and `-1`, or not? I also assume this was copied from the simple range analysis library?

MathiasVP replied:
Indeed this was copied from simple range analysis. Why should we have `0` and `-1` here? AFAIK, the bounds should be inclusive.

jketema replied:
Ah, the standard says: "If the source type is bool, the value false is converted to zero and the value true is converted to one." So what you added seems correct.

geoffw0 commented:
Why *wouldn't* we want to deduce that `ull <= UINT_MAX` in the above example?

jketema commented on cpp/ql/test/library-tests/ir/range-analysis/SimpleRangeAnalysis_tests.cpp:
```diff
@@ -759,9 +759,9 @@ unsigned long mult_overflow() {
     range(ui); // $ range=>=10
-    range((unsigned long)ui); // $ range=>=10
-    unsigned long result = (unsigned long)ui * ui; // $ overflow=+
-    range(result); // $ MISSING: range=>=100
```
Was the annotation wrong here, or does adding the upperbound here magically fix things?

MathiasVP replied:
Adding the upper-bound here magically fix things :) Because we can prove that it doesn't overflow we're now able to report a bound.

jketema commented on cpp/ql/test/library-tests/ir/range-analysis/SimpleRangeAnalysis_tests.cpp:
```diff
@@ -888,7 +888,7 @@ void notequal_variations(short n, float f) {
 
   if (n >= 5) {
-    if (2 * n - 10 == 0) { // $ overflow=+
+    if (2 * n - 10 == 0) { // no overflow
```
Why does this change?

MathiasVP replied:
`n` is a short, and `2` is a int. So `n` gets promoted to an int, and we infer that `n <= SHRT_MAX` which fits well within an integer (which is the resulting type of `2 * n`)

jketema reviewed (commented):
Some small comments, just to further my understanding of the code.

MathiasVP commented:
> Why _wouldn't_ we want to deduce that `ull <= UINT_MAX` in the above example?

Because it's not a very precise bound, and many users have been confused by false positives caused by such very-large-but-sound bounds obtained from type-information only. For example, [this change](https://github.com/github/codeql/pull/7386) was made because we wanted to be able to distinguish between precise bounds found by bounds from guards, and less precise bounds from type-information, for a high precision version of the `cpp/overrunning-write` query.

And then there's this beauty from an external contributor who had a similar problem: https://github.com/github/codeql/blob/main/cpp/ql/src/experimental/Security/CWE/CWE-561/FindIncorrectlyUsedSwitch.ql#L21

jketema commented:
> > Why _wouldn't_ we want to deduce that `ull <= UINT_MAX` in the above example?
> 
> Because it's not a very precise bound, and many users have been confused by false positives caused by such very-large-but-sound bounds obtained from type-information only. For example, [this change](https://github.com/github/codeql/pull/7386) was made because we wanted to be able to distinguish between precise bounds found by bounds from guards, and less precise bounds from type-information, for a high precision version of the `cpp/overrunning-write` query.
> 
> And then there's this beauty from an external contributor who had a similar problem: https://github.com/github/codeql/blob/main/cpp/ql/src/experimental/Security/CWE/CWE-561/FindIncorrectlyUsedSwitch.ql#L21

Not as part of this PR, but as each bound comes with a reason, we might want to explore whether we can add a new reason for the type bounds that are introduced here.

jketema commented:
The DCA alert changes do not make any kind of sense to me. Any clue what was going on there?

MathiasVP commented:
- The `cpp/uncontrolled-arithmetic` ones in Samate definitely LGTM since they're all in functions annotated as TPs.
  - I _think_ the lost result on `abseil-cpp-linux` is correct assuming `MultType` resolves to `uint128`. If it resolves to `uint64_t` I'm less sure. Will need to check that locally, I think.
- Most of the results on `cpp/constant-comparison` for `ImageMagick/ImageMagick` also LGTM, but I'm still not sure about all of them.
- The `cpp/integer-multiplication-cast-to-long` results LGTM
- The `cpp/overrunning-write` results looks like the result of inferring upper bounds purely from the type. That's also how the query behaves on `main`, though, so that's not an issue.
- The `cpp/unbounded-write` results LGTM. All the lost results are because we correctly infer a smaller upper bound based on a type promotion.
- Same story for `cpp/uncontrolled-allocation-size`.
- The new result on `cpp/very-likely-overrunning-write` for lubomyr__bochs is a TP, I think: `pname` has space for 6 chars, but `port%d` can have length `strlen("port")` plus the size of `max(i + 1)`, and `i` is upper bounded by `hub.n_ports` (which is an unsigned char). So `max(i + 1) = 256`. So it does seem like the `sprintf` requires 4 + 3 + 1 = 8 (for the null terminator) to be safe.
  - The new result on `vim` is absolutely horrible 😂. I can't parse that `sprintf` in my head, but I'm very sure the new results comes because we're able to infer that `key_name[0]` and `key_name[1]` can each be at most `CHAR_MAX`.

jketema commented:
Thanks. I would be fine with having this merged, assuming DCA against `main` for both the nightly suite and the MCTV suite look fine. I'm not sure what @geoffw0 thinks?

MathiasVP commented:
Good point. I actually didn't test this against `main` (I only tested it against https://github.com/github/codeql/pull/12505), but I'll start such a run right away.

MathiasVP commented:
Uh oh. Since we now infer many many more constant bounds the `cpp/constant-array-overflow` query suddenly has a lot more results now. There's also two new FPs on `cpp/invalid-pointer-deref` with an alert message that reads very much like we're getting a type-based bound:

```
This read might be out of bounds, as the pointer might be equal to call to malloc + position + 65534.
```

I think that's another good reason to do what Jeroen said here:

> > > Why _wouldn't_ we want to deduce that `ull <= UINT_MAX` in the above example?
> > 
> > 
> > Because it's not a very precise bound, and many users have been confused by false positives caused by such very-large-but-sound bounds obtained from type-information only. For example, [this change](https://github.com/github/codeql/pull/7386) was made because we wanted to be able to distinguish between precise bounds found by bounds from guards, and less precise bounds from type-information, for a high precision version of the `cpp/overrunning-write` query.
> > And then there's this beauty from an external contributor who had a similar problem: https://github.com/github/codeql/blob/main/cpp/ql/src/experimental/Security/CWE/CWE-561/FindIncorrectlyUsedSwitch.ql#L21
> 
> Not as part of this PR, but as each bound comes with a reason, we might want to explore whether we can add a new reason for the type bounds that are introduced here.

But since this is causing FPs on `cpp/invalid-pointer-deref` it may be worth including the necessary changes to `SemReason` so that we can exclude these cases in the query.

MathiasVP commented:
I've pushed two temporary commits to this PR: https://github.com/github/codeql/pull/13783/commits/54b903c60d5075fc8d3feb1e9adc6e64aaced584 and https://github.com/github/codeql/pull/13783/commits/59b7d3ef0c745bfbbbdaf0a754470de54afb2c9b. 

Together, these filter out type-based bounds from the two MCTV queries that DCA showed had new FPs as a result of these changes. If DCA comes back happy I'll move those commits commits to another PR so that we can merge that one before continuing with this PR.

geoffw0 commented:
> I'm not sure what @geoffw0 thinks?

Sorry I didn't see the notification for this until now.  I'll have a quick look at the DCA alert differences on the latest run, particularly for `cpp/overrunning-write` and `cpp/unbounded-write`...

MathiasVP commented:
> > I'm not sure what @geoffw0 thinks?
> 
> Sorry I didn't see the notification for this until now. I'll have a quick look at the DCA alert differences on the latest run, particularly for `cpp/overrunning-write` and `cpp/unbounded-write`...

Please don't do so yet. I'm still fixing things. I'll move this to a draft to make this explicit.

MathiasVP commented:
My long series of TEMP commits seems to have removed all the type-related FPs, and we're now left with four new results. I _thought_ these were FPs similar to the other ones I've fixed in the earlier commits, but these turn out to be genuine (I mean ... they're still a FP wrt. the query, but the range analyses reported are all correct).

For example, one of the four new alerts is on [this line](https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/java.desktop/share/native/libsplashscreen/libpng/pngread.c#L2497) where we report that the `back_r` used as the index into `png_sRGB_table` may have the value `65535`, while the array is only 255 entries long (giving us an "off-by-65280" alert because `65535 - 255 = 65280`). This very much looked like a type-bound was still somehow getting through to the query even though I've removed type-based bounds from the query in https://github.com/github/codeql/pull/13783/commits/a7194c9d524b8f3704497d36038a0dd36c4d81a9 and https://github.com/github/codeql/pull/13783/commits/03b72eb6d1650a41ff39030b952ffa2ee076c13e.

However! After a long and painful debugging session I found this [line](https://github.com/openjdk/jdk/blob/da75f3c4ad5bdf25167a3ed80e51f567ab3dbd01/src/java.desktop/share/native/libsplashscreen/libpng/pngread.c#L2170) which is the source of this reported `65535`. So the result is "correct", although the path that would allow for this overflow is infeasible.

TLDR: It seems like things work now 👍. I'll cleanup this PR and take it out of draft later today.

MathiasVP commented:
I've moved all the TEMP commits to a fresh PR: https://github.com/github/codeql/pull/13880. Once that PR is merged I can rebase this PR and pull this out of draft 🎉

MathiasVP commented:
DCA looks good. We have the same set of results as the very first DCA result (all of which we verified). There was a slowdown on Wireshark which I could reproduce locally, and a subsequent DCA run didn't show any slowdown either. So I think this PR is good to go!

jketema commented on cpp/ql/lib/semmle/code/cpp/rangeanalysis/new/internal/semantic/analysis/RangeAnalysisConstantSpecific.qll:
```diff
@@ -74,7 +74,10 @@ module CppLangImplConstant implements LangSig<FloatDelta> {
-  predicate hasConstantBound(SemExpr e, float bound, boolean upper) { none() }
+  predicate hasConstantBound(SemExpr e, float bound, boolean upper, SemReason reason) {
+    semHasConstantBoundConstantSpecific(e, bound, upper) and
+    reason instanceof SemTypeReason
```
I've trouble following this. Why os this restricted to `SemTypeReason` only?

MathiasVP replied:
Because `semHasConstantBoundConstantSpecific` only infers bounds based on type information. It's probably possible to push this condition into `semHasConstantBoundConstantSpecific`, but that then requires that `semHasConstantBoundConstantSpecific` knows about the `SemReason` type which it currently doesn't since that predicate is defined outside the `CppLangImplConstant` module (which instantiates those types).

Does that make sense?

jketema replied:
So zero bounds are not constant?

MathiasVP replied:
`semHasConstantBoundConstantSpecific` just forwards to `hasConstantBoundConstantSpecific` (which is the predicate added [here](https://github.com/github/codeql/pull/13783/commits/050a563580aade23ed1d2e2b755921953c9c38a3#diff-f66e5037244aec24348feede9c60d2584ef91c6b045466f21c6f307d72107008R461)). The only thing that adds is the type bounds inferred from type conversions.

jketema replied:
We briefly discussed this face-to-face. This is the correct thing to do here because as Mathias writes:

> `semHasConstantBoundConstantSpecific` only infers bounds based on type information

We could extend this later and push down the `reason` argument, but that is not of immediate concern here.

jketema reviewed (commented):
One question, otherwise LGTM.
